docs/source/tutorials_source/package/tutorial_moco_memory_bank.py

```diff
@@ -208,52 +215,72 @@
 # .. note:: We use a split batch norm to simulate multi-gpu behaviour. Combined
 #   with the use of batch shuffling, this prevents the model from communicating
 #   through the batch norm layers.
-
 class MocoModel(pl.LightningModule):
     def __init__(self):
         super().__init__()
         
         # create a ResNet backbone and remove the classification head
         resnet = lightly.models.ResNetGenerator('resnet-18', 1, num_splits=8)
-        backbone = nn.Sequential(
+        self.backbone = nn.Sequential(
             *list(resnet.children())[:-1],
             nn.AdaptiveAvgPool2d(1),
         )
 
-        # create a moco based on ResNet
-        self.resnet_moco = \
-            lightly.models.MoCo(backbone, num_ftrs=512, m=0.99, batch_shuffle=True)
+        # create a moco model based on ResNet
+        self.projection_head = MoCoProjectionHead(512, 512, 128)
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+        deactivate_requires_grad(self.backbone_momentum)
+        deactivate_requires_grad(self.projection_head_momentum)
 
         # create our loss with the optional memory bank
         self.criterion = lightly.loss.NTXentLoss(
             temperature=0.1,
             memory_bank_size=memory_bank_size)
 
-    def forward(self, x):
-        self.resnet_moco(x)
+    def training_step(self, batch, batch_idx):
+        (x_q, x_k), _, _ = batch
 
-    # We provide a helper method to log weights in tensorboard
-    # which is useful for debugging.
-    def custom_histogram_weights(self):
-        for name, params in self.named_parameters():
-            self.logger.experiment.add_histogram(
-                name, params, self.current_epoch)
+        # update momentum
+        update_momentum(self.backbone, self.backbone_momentum, 0.99)
+        update_momentum(
+            self.projection_head, self.projection_head_momentum, 0.99
+        )
 
-    def training_step(self, batch, batch_idx):
-        (x0, x1), _, _ = batch
-        y0, y1 = self.resnet_moco(x0, x1)
-        loss = self.criterion(y0, y1)
-        self.log('train_loss_ssl', loss)
+        # get queries
+        q = self.backbone(x_q).flatten(start_dim=1)
+        q = self.projection_head(q)
+
+        # get keys
+        k, shuffle = batch_shuffle(x_k)
+        k = self.backbone_momentum(k).flatten(start_dim=1)
+        k = self.projection_head_momentum(k)
+        k = batch_unshuffle(k, shuffle)
+
+        loss = self.criterion(q, k)
```

Comment:
Do I assume correctly, that this q, k, are the same ones as used in this image?
https://user-images.githubusercontent.com/11435359/71603927-0ca98d00-2b14-11ea-9fd8-10d984a2de45.png

If yes, could you add this image to the tutorial? I think it would make it easier to understand these steps.

Reply:
Maybe we should discuss whether we want to add these images to the tutorials in general (SimCLR, SimSiam, ...). I feel that the tutorials are already quite long and would let people have a look at the paper if they are interested.

Reply:
Ok, then let's leave them away for the moment.